vmm/src/vstate.rs

```diff
@@ -280,6 +292,84 @@ pub struct Vcpu {
 }
 
 impl Vcpu {
+    thread_local!(static TLS_VCPU_PTR: VcpuCell = Cell::new(None));
+
+    /// Associates `self` with the current thread.
+    ///
+    /// Is a prerequisite to succesfully run `run_on_thread_local_vcpu()` on the current thread.
+    /// Will fail if there is already a `Vcpu` present in this thread's TLS.
+    fn init_thread_local_data(&mut self) -> Result<()> {
+        Self::TLS_VCPU_PTR.with(|cell: &VcpuCell| {
+            if cell.get().is_some() {
+                return Err(Error::VcpuTlsInit);
+            }
+            cell.set(Some(self as *mut Vcpu));
+            Ok(())
+        })
+    }
+
+    /// Deassociates `self` from the current thread.
+    ///
+    /// Should be called if the current `self` had called `init_thread_local_data()` and
+    /// now needs to move to a different thread.
+    ///
+    /// Fails if `self` was not previously associated with the current thread.
+    fn reset_thread_local_data(&mut self) -> Result<()> {
+        // Best-effort to clean up TLS. If the `Vcpu` was moved to another thread
+        // _before_ running this, then there is nothing we can do.
+        let mut tls_contains_self = false;
+        Self::TLS_VCPU_PTR.with(|cell: &VcpuCell| {
+            if let Some(vcpu_ptr) = cell.get() {
+                if vcpu_ptr == self as *mut Vcpu {
+                    tls_contains_self = true;
+                }
+            }
+        });
+
+        if tls_contains_self {
+            Self::TLS_VCPU_PTR.with(|cell: &VcpuCell| cell.take());
+            Ok(())
+        } else {
+            Err(Error::VcpuTlsNotPresent)
+        }
+    }
+
+    /// Runs `func` for the `Vcpu` associated with the current thread.
+    ///
+    /// It requires that `init_thread_local_data()` was run on this thread.
+    ///
+    /// Fails if there is no `Vcpu` associated with the current thread.
+    fn run_on_thread_local_vcpu<F>(func: F) -> Result<()>
```

Comment:
I guess we can remove the `vcpu` from the name, since it's implied by this being a method in `Vcpu`. Also, we might want to mark this as `unsafe`, since it seems to allow the aliasing of a `&mut Vcpu` (we can have a mutable reference, and effectively get a second one in the closure we pass to this method).

Reply:
This isn't quite doing RAII because the resource is not owned by the vcpu, but by the thread, which the vcpu isn't tied to. I guess the closest thing to actual RAII we could do here is put vcpu objects in a wrapper struct (that does the tls init in its constructor, and the reset in its destructor instead of the vcpu), and also make sure the wrapper cannot be moved to another thread by implementing `!Send`.